packages/aws-cdk/lib/api/hotswap/cloudwatch-logs.ts

```diff
@@ -0,0 +1,97 @@
+  const implicitMap = new Map<string, string>();
+  //
+  // resolve the environment, so we can substitute things like AWS::Region in CFN expressions
+  const resolvedEnv = await sdkProvider.resolveEnvironment(stackArtifact.environment);
+  // create a new SDK using the CLI credentials, because the default one will not work for new-style synthesis -
+  // it assumes the bootstrap deploy Role, which doesn't have permissions to update Lambda functions
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForWriting);
+
+  // add the sdk to the hotswapLogMonitor so that it also has the
+  // correct credentials to monitor CloudWatch log groups
+  hotswapLogMonitor.addSdk(sdk);
+  // The current resources of the Stack.
+  // We need them to figure out the physical name of a resource in case it wasn't specified by the user.
+  // We fetch it lazily, to save a service call, in case all hotswapped resources have their physical names set.
+  const listStackResources = new LazyListStackResources(sdk, stackArtifact.stackName);
+  const evaluateCfnTemplate = new EvaluateCloudFormationTemplate({
+    stackArtifact,
+    parameters: assetParams,
+    account: resolvedEnv.account,
+    region: resolvedEnv.region,
+    partition: (await sdk.currentAccount()).partition,
+    urlSuffix: sdk.getEndpointSuffix,
+    listStackResources,
+  });
+
+  const resources = await listStackResources.listStackResources();
+
+  // do a first pass at identifying all log groups
+  for (const resource of resources) {
+    if (resource.ResourceType === 'AWS::Logs::LogGroup') {
+      logGroupLogicalIds.add(resource.LogicalResourceId);
+      // track all the resources that will create a log group if one
+      // is not provided
+    } else if (implicitTypes.includes(resource.ResourceType)) {
+      const pathPart = resource.ResourceType.split('::')[1].toLowerCase();
+      implicitMap.set(resource.LogicalResourceId, pathPart);
+    }
+  }
+
+  // For each CloudWatch Log group in the template, find all resources that
+  // reference that log group and if it matches once of the included types, add it.
+  for (const id of logGroupLogicalIds) {
+    const resourcesReferencingLogGroup = evaluateCfnTemplate.findReferencesTo(id);
+    for (const r of resourcesReferencingLogGroup) {
+      if (includedTypes.includes(r.Type)) {
+        const groupName = await evaluateCfnTemplate.findPhysicalNameFor(id);
+        if (groupName) {
+          logGroupNames.push(groupName);
+        }
+        // we are explicitely creating a cloudwatch log group for this resource
+        // so add that log group and remove the resource from the list of resources
+        // that might have an implicitely created log group
+      } else if (implicitTypes.includes(r.Type)) {
+        implicitMap.delete(r.LogicalId);
+        const groupName = await evaluateCfnTemplate.findPhysicalNameFor(id);
+        if (groupName) {
+          logGroupNames.push(groupName);
+        }
```

Comment:
We really should get rid of this duplication between the two branches of the `if`.